DIPs/DIP4242.md

```diff
@@ -0,0 +1,260 @@
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+As of v2.095.0, there is no easy way to write a non-templated function that accepts
+a delegate parameter and allow a wide range of attributes.
+
+For example, when writing a `@safe` function, one is faced with a restrictive choice:
+either marks the delegate as `@safe`, and force the caller to use `@trusted` in some occasions,
+or avoid marking the function itself `@safe`, and not be callable from `@safe` code.
+In general, the former seems to be the prefered approach, as it makes the most sense.
+
+However, this choice is much less obvious for other attributes: forcing `nothrow`
+or `pure`ness is a less-accepted practice, let alone how restrictive forcing `@nogc` is.
+
+This problem can be seen in many widely used library, such as Vibe.d's delegate-accepting
+[`requestHTTP`](https://vibed.org/api/vibe.http.client/requestHTTP),
+or even druntime's [`Throwable.toString`](https://github.com/dlang/druntime/blob/d97ec4093b108dc2fa95f1fa04b1114e6e0611f8/src/object.d#L2020-L2026).
+It is also commonly seen when implementing `opApply`, as one usually has to choose between working type-inferencex
+(e.g. `foreach (varname; container)` as opposed to `foreach (type varname; container)`),
+which only works if the delegate type is known and not templated, or supporting multiple
+attributes, which is done by templating the delegate type.
+
+This proposal adds the ability to express the common dependency between a delegate's
+attributes and a function's attribute.
+
+## Prior Work
+
+[DIP1032](DIP1032.md) has proposed to always have the delegate parameters take the
+attributes of the function that receives it.
+However, this does not address the problem presented here: instead, it forces
+the user to require from the caller the attributes it supports.
+
+Doing so usually leads to attributes being faked, or the routine being unusable:
+for example, iterating over a collection might be a simple operation which is
+`pure` and `@nogc`, but requiring the delegate to also be is too limitating.
+
+## Description
+
+### Grammar
+
+The following changes to the grammar are proposed:
+```diff
+ StorageClass:
+     LinkageAttribute
+     AlignAttribute
+     deprecated
+     enum
+     static
+     extern
+     abstract
+     final
+     override
+     synchronized
+     auto
+     scope
+     const
+     immutable
+     inout
+     shared
+     __gshared
+     Property
+     nothrow
++    nothrow AttributeDeclDef
+     pure
++    pure AttributeDeclDef
+     ref
+ 
+ AtAttribute:
+     @ disable
+     @ nogc
++    @ nogc AttributeDeclDef
+     @ live
+     Property
+     @ safe
++    @ safe AttributeDeclDef
+     @ system
+     @ trusted
+     UserDefinedAttribute
+ 
++AttributeDeclDef:
++    ( * )
```

Comment:
The meaning of `*` is never mentioned AFAICT.
I guess it's a short hand notation for all parameters?

Reply:
Indeed! Good catch, will add a paragraph.